File: assignments/SaveDataProject/app/src/main/java/com/example/savedataproject/MainActivity.java
package com.example.savedataproject;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private EditText nameInput;
    private LinearLayout namesContainer;
    private NamesViewModel namesViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        nameInput = findViewById(R.id.nameInput);
        namesContainer = findViewById(R.id.namesContainer);

        namesViewModel = new ViewModelProvider(this).get(NamesViewModel.class);

        Button saveButton = findViewById(R.id.saveButton);
        saveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveName();
            }
        });

        displayNames(namesViewModel.getSavedNames());
    }

    private void saveName() {
        String enteredName = nameInput.getText().toString().trim();

        if (!enteredName.isEmpty()) {
            namesViewModel.addName(enteredName);

            displayNames(namesViewModel.getSavedNames());

            nameInput.getText().clear();
        } else {

            nameInput.setError("No name entered");
        }
    }

    private void displayNames(List<String> names) {
        namesContainer.removeAllViews();

        if (names.isEmpty()) {
            TextView noNameTextView = new TextView(this);
            noNameTextView.setText("No name entered");
            namesContainer.addView(noNameTextView);
        } else {

            for (String name : names) {
                TextView textView = new TextView(this);
                textView.setText(name);

                namesContainer.addView(textView);
            }
        }
    }
}

File: assignments/LifeCycleAware/app/src/main/java/com/example/lifecycleaware/MainActivity.kt
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.Observer
import com.example.lifecycleaware.R

class MainActivity : AppCompatActivity() {

    private lateinit var viewModel: MyViewModel
    private lateinit var observer: MyObserver

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        viewModel = ViewModelProvider(this).get(MyViewModel::class.java)

        observer = MyObserver()
        lifecycle.addObserver(observer)

        viewModel.getEventsLiveData().observe(this, Observer { events ->
            for (event in events) {
                Log.d("ViewModelEvents", "Event: $event")
            }
        })

        viewModel.addEvent("onCreate")
    }
}
